cloud/vsphere/machineactuator.go

```diff
@@ -510,69 +470,113 @@ func (vc *VsphereClient) Update(cluster *clusterv1.Cluster, goalMachine *cluster
-			goalMachineKubeletVersion != goalMachine.Annotations[KubeletVersionAnnotationKey] {
-			glog.Infof("Doing in-place upgrade for master from %s to %s", goalMachine.Annotations[ControlPlaneVersionAnnotationKey], goalMachineControlPlaneVersion)
+		if vc.needsMasterUpdate(goalMachine) {
+			glog.Infof("Doing in-place upgrade for master from v%s to v%s", goalMachine.Annotations[ControlPlaneVersionAnnotationKey], goalMachine.Spec.Versions.ControlPlane)
 			err := vc.updateMasterInPlace(goalMachine)
 			if err != nil {
-				glog.Errorf("Master inplace update failed: %+v", err)
+				glog.Errorf("Master in-place upgrade failed: %+v", err)
+				return err
 			}
 		} else {
-			glog.Info("UPDATE TYPE NOT SUPPORTED")
-			return nil
+			glog.Info("UNSUPPORTED MASTER UPDATE.")
 		}
 	} else {
-		glog.Info("NODE UPDATES NOT SUPPORTED")
-		return nil
+		if vc.needsNodeUpdate(goalMachine) {
+			// Node upgrades
+			if err := vc.updateNode(cluster, goalMachine); err != nil {
+				glog.Errorf("Node %s update failed: %+v", goalMachine.ObjectMeta.Name, err)
+				return err
+			}
+		} else {
+			glog.Info("UNSUPPORTED NODE UPDATE.")
+		}
 	}
 
 	return nil
 }
 
-// Assumes that update is needed.
-// For now support only K8s control plane upgrades.
-func (vc *VsphereClient) updateMasterInPlace(goalMachine *clusterv1.Machine) error {
-	goalMachineControlPlaneVersion := goalMachine.Spec.Versions.ControlPlane
-	goalMachineKubeletVersion := goalMachine.Spec.Versions.Kubelet
+func (vc *VsphereClient) needsControlPlaneUpdate(machine *clusterv1.Machine) bool {
+	return machine.Spec.Versions.ControlPlane != machine.Annotations[ControlPlaneVersionAnnotationKey]
+}
 
-	currentControlPlaneVersion := goalMachine.Annotations[ControlPlaneVersionAnnotationKey]
-	currentKubeletVersion := goalMachine.Annotations[KubeletVersionAnnotationKey]
+func (vc *VsphereClient) needsKubeletUpdate(machine *clusterv1.Machine) bool {
+	return machine.Spec.Versions.Kubelet != machine.Annotations[KubeletVersionAnnotationKey]
+}
 
-	// Control plane upgrade
-	if goalMachineControlPlaneVersion != currentControlPlaneVersion {
+// Returns true if the node is needed to be upgraded.
+func (vc *VsphereClient) needsNodeUpdate(machine *clusterv1.Machine) bool {
+	return !util.IsMaster(machine) &&
+		vc.needsKubeletUpdate(machine)
+}
+
+// Returns true if the master is needed to be upgraded.
+func (vc *VsphereClient) needsMasterUpdate(machine *clusterv1.Machine) bool {
+	return util.IsMaster(machine) &&
+		vc.needsControlPlaneUpdate(machine)
+	// TODO: we should support kubelet upgrades here as well.
+}
+
+func (vc *VsphereClient) updateKubelet(machine *clusterv1.Machine) error {
+	if vc.needsKubeletUpdate(machine) {
+		// Kubelet packages are versioned 1.10.1-00 and so on.
+		kubeletAptVersion := machine.Spec.Versions.Kubelet + "-00"
+		cmd := fmt.Sprintf("sudo apt-get install kubelet=%s", kubeletAptVersion)
+		if _, err := vc.remoteSshCommand(machine, cmd, "~/.ssh/vsphere_tmp", "ubuntu"); err != nil {
```

Comment:
is there any reason to not fix the privateKeyPath and sshUser?
seems like within the vsphere provider, these values will always be the same.

Reply:
Not really, just legacy. This will be easier to do as part of https://github.com/kubernetes-sigs/cluster-api/issues/167 so I'm leaving it as is for now.